### PR TITLE
Adding AWS X-Ray tracer with 5% default sampling rate.

### PR DESCRIPTION
**Pull request by luluzhao**
Signed-off-by: Lulu Zhao <luluzhao@amazon.com>

Description: The implementation of X-Ray to be one of the tracing extension inside Envoy proxy, which with 5% sampling rate for spans. 
Testing: Unit test, Integration test
Docs Changes: Need a separate section for X-Ray among with other tracers.
[#Issue] Related to: [5644](https://github.com/envoyproxy/envoy/issues/5644)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by dnoe** [comment]
Make sure you have done format fix as described here: https://github.com/envoyproxy/envoy/blob/master/support/README.md#fixing-format-problems

The format check is failing.

**Comment by jmarantz** [comment]
Once the formatting fixes are in (in particular using the Envoy abstractions for time, mutex, and condvar) I wonder if one of the people involved in Hystrix would be an appropriate initial reviewer, maybe @trabetti or @rgs1 ?

**Comment by luluzhao** [comment]
I will address the formatting issue later and thank you so much for taking a look at it.

File: source/common/json/config_schemas.cc [hunk 1]
```diff
@@ -1321,6 +1341,7 @@ const std::string Json::Schema::TOP_LEVEL_CONFIG_SCHEMA(R"EOF(
               "driver" : {
                 "type" : "object",
                 "oneOf" : [
+                  {"$ref" : "#/definitions/xray_driver"},
                   {"$ref" : "#/definitions/lightstep_driver"},
                   {"$ref" : "#/definitions/zipkin_driver"}
                 ]
```
Comment (venilnoronha) [inline source/common/json/config_schemas.cc]: Changes to this file are no longer required as this is for v1 only.

**Comment by jmarantz** [comment]
cool; note that most of the formatting issues can be fixed automatically, though not the mutex/timer abstractions.

File: source/extensions/tracers/well_known_names.h [hunk 2]
```diff
@@ -21,6 +21,8 @@ class TracerNameValues {
   const std::string DynamicOt = "envoy.dynamic.ot";
   // Datadog tracer
   const std::string Datadog = "envoy.tracers.datadog";
+  // XRay tracer
+  const std::string XRay = "envoy.xray";
```
Comment (venilnoronha) [inline source/extensions/tracers/well_known_names.h]: The naming convention for tracers is `envoy.tracers.name`. Could you change this to `envoy.tracers.xray`?

File: .gitignore [hunk 3]
```diff
@@ -28,3 +28,6 @@ TAGS
 .vimrc
 .vs
 .vscode
+.DS_Store
+.project/
+.idea/
```
Comment (venilnoronha) [inline .gitignore]: Order by alpha.

File: source/common/json/config_schemas.cc (hunk 1 above)
Comment (luluzhao) [reply source/common/json/config_schemas.cc]: I thought the next release will be v1.10, is that correct? If it is not required, I will delete this line. Thanks.

File: source/extensions/tracers/well_known_names.h (hunk 2 above)
Comment (luluzhao) [reply source/extensions/tracers/well_known_names.h]: Sure thing. Will change it to `envoy.tracers.xray`.

File: source/common/json/config_schemas.cc (hunk 1 above)
Comment (venilnoronha) [reply source/common/json/config_schemas.cc]: JSON schemas were used previously. We now rely on Protobufs.

**Comment by mattklein123** [comment]
Agreed let's please break this up and make sure we can get a small PR in with the correct style/tests, etc. Thank you.

**Comment by luluzhao** [comment]
Thank you all for the suggestions. I will break it up to several small PRs. In order to start implementing X-Ray tracer, I was looking at how other tracers got implemented like Zikpin and try to do the same style like it. For example, I didn't use "= default" for the copy constructors is because of [this](https://github.com/envoyproxy/envoy/blob/master/source/extensions/tracers/zipkin/zipkin_core_types.cc#L18). It got published so I thought I should do similar style. I will read the link about Envoy style coding and correct the style.

**Pull request closed without merging**